### PR TITLE
Fix crash in symbolkeywriting.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems?id=249521&fullScreen=false&_a=edit

**Comment by CyrusNajmabadi** [comment]
test vsi please

**Comment by Pilchie** [comment]
@CyrusNajmabadi Can you rebase/submit this against `dev15-preview-4`?

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb [hunk 1]
```diff
@@ -1857,12 +1857,35 @@ class Program
...
                 state.SendTypeChars("(")
 
-                Await state.WaitForAsynchronousOperationsAsync().ConfigureAwait(True)
+                Await state.WaitForAsynchronousOperationsAsync()
                 state.AssertMatchesTextStartingAtLine(7, "new[] { new { x = 1 } }.ToArray(")
             End Using
         End Function
 
         <WpfFact, Trait(Traits.Feature, Traits.Features.Completion)>
+        Public Async Function TestRecursiveGenericSymbolKey() As Task
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb]: Shouldn't this be a SymbolKey test instead of a completion test?

Comment (jasonmalinowski) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb]: I'd expect another test anyways that confirms we can actually resolve the SymbolKey too.

File: src/Workspaces/Core/Portable/SymbolKey/SymbolKey.SymbolKeyWriter.cs [hunk 2]
```diff
@@ -165,7 +165,29 @@ private void WriteSymbolKey(ISymbol symbol, bool first)
 
                 if (!shouldWriteOrdinal)
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/SymbolKey/SymbolKey.SymbolKeyWriter.cs]: I don't get this variable...isn't it writing an ordinal? Or is "ordinal" and the "id" separate things here?

File: src/Workspaces/Core/Portable/SymbolKey/SymbolKey.SymbolKeyWriter.cs [hunk 3]
```diff
@@ -165,7 +165,29 @@ private void WriteSymbolKey(ISymbol symbol, bool first)
...
+                    //
+                    // The recursion is prevented because when we hit 'Foo' we mark that
+                    // we're writing out a signature.  And, in signature mode we only write
+                    // out the ordinal for 'Z' without recursing.  However, even though
+                    // we prevent the recursion, we still hit List<Z> twice.  After writing
+                    // the innermost one out, we'll give it a reference ID.  When we
+                    // then hit the outermost one, we want to just reuse that one.
+                    if (!_symbolToId.ContainsKey(symbol))
+                    {
+                        _symbolToId.Add(symbol, id);
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/SymbolKey/SymbolKey.SymbolKeyWriter.cs]: But we still generated a number and wrote that out. Is that bad?

**Comment by jasonmalinowski** [comment]
The code you're touching seems sufficiently confusing to me to understand if this is the correct fix. It seems a _lot_ comments are needed in that function to provide the context needed.

Further, this should be tested by the SymbolKey tests against the API directly, not at the completion layer. That would give us more clear coverage that you can resolve one of these keys too, which we might be missing right now.

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb]: I'll add another test.

File: src/Workspaces/Core/Portable/SymbolKey/SymbolKey.SymbolKeyWriter.cs [hunk 4]
```diff
@@ -184,12 +183,24 @@ private void WriteSymbolKey(ISymbol symbol, bool first)
...
                     {
+                        // Haven't hit this symbol before, write out its fresh ID.
                         _symbolToId.Add(symbol, id);
                     }
                 }
 
+                // Now write out the ID for this symbol so that any future hits of it can 
+                // write out a reference to it instead.
+                WriteInteger(id);
+
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/SymbolKey/SymbolKey.SymbolKeyWriter.cs]: I find this version of the change much easier to read. Thanks!

**Comment by Pilchie** [comment]
:+1: for the change, but needs to be against `dev14-preview-4`

**Pull request closed without merging**